File: py110-intermediate/lesson_3/21.py
import random
import os


PLAYER = 'Player'
DEALER = 'Dealer'
'''
DECK = [
        ['S', '2'], ['S', '3'], ['S', '4'], ['S', '5'], ['S', '6'],
        ['S', '7'], ['S', '8'], ['S', '9'], ['S', '10'], ['S', 'J'],
        ['S', 'Q'], ['S', 'K'], ['S', 'A'],
        ['H', '2'], ['H', '3'], ['H', '4'], ['H', '5'], ['H', '6'],
        ['H', '7'], ['H', '8'], ['H', '9'], ['H', '10'], ['H', 'J'],
        ['H', 'Q'], ['H', 'K'], ['H', 'A'],
        ['C', '2'], ['C', '3'], ['C', '4'], ['C', '5'], ['C', '6'],
        ['C', '7'], ['C', '8'], ['C', '9'], ['C', '10'], ['C', 'J'],
        ['C', 'Q'], ['C', 'K'], ['C', 'A'],
        ['D', '2'], ['D', '3'], ['D', '4'], ['D', '5'], ['D', '6'],
        ['D', '7'], ['D', '8'], ['D', '9'], ['D', '10'], ['D', 'J'],
        ['D', 'Q'], ['D', 'K'], ['D', 'A'],
    ]
'''
SUITS = ('Hearts', 'Diamonds', 'Spades', 'Clubs')
VALUES = ('2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A')

DECK = [[value, suit] for value in VALUES for suit in SUITS]


def terminal_clear():
    if os.name == 'posix':
        os.system('clear')
    else:
        os.system('cls')


def prompt(message):
    print(f"===> {message}")


def deal_cards():
    hand = [random.choice(DECK)]
    
    # deal 2nd card, re-deal if same as 1st
    while True:
        dealt = random.choice(DECK)

        if dealt not in hand:
            hand.append(dealt)

            return hand


def display_initial_hands(player_hand, dealer_hand):
    prompt(f"Dealer's hand: {' of '.join(dealer_hand[0])} and ?")
    prompt(f"Player's hand: {' of '.join(player_hand[0])} and {' of '.join(player_hand[1])}")
    print()


def display_current_hands(cards_in_hand):
    return ', '.join([' of '.join(card) for card in cards_in_hand])


def hit(hand):
    while True:
        dealt = random.choice(DECK)

        if dealt not in hand:
            hand.append(dealt)

            return
        

def dealer_stays(dealer_hand):
    card_total = total(dealer_hand)

    if card_total >= 17:
        return True


def busted(cards_in_hand):
    current_total = total(cards_in_hand)

    if current_total > 21:
        return True
    
    return False


def player_turn(player_hand):
    while True:
        if busted(player_hand):
            break

        prompt(f"TOTAL: {total(player_hand)}")
        prompt("hit (h) or stay (s)?")
        answer = input().lower()

        if answer == 's':
            break
        elif answer == 'h':
            hit(player_hand)
            prompt(f"PLAYER'S hand: {display_current_hands(player_hand)}")
        else:
            prompt("Try again you silly goose!")
    
    if busted(player_hand):
        prompt(f"You BUSTED! Card total: {total(player_hand)}\n")
    else:
        prompt(f"You chose to stay! Card total: {total(player_hand)}\n")


def dealer_turn(dealer_hand):
    while True:
        dealer_total = total(dealer_hand)
        prompt(f"DEALER'S hand: {display_current_hands(dealer_hand)}")
        prompt(f"TOTAL: {dealer_total}")

        if busted(dealer_hand):
            prompt(f"DEALER BUSTED!  Card total: {dealer_total}\n")
            break
        elif dealer_stays(dealer_hand):
            prompt(f"DEALER stays.  Card total: {dealer_total}\n")
            break

        hit(dealer_hand)


def total(cards_in_hand):
    values = [card[0] for card in cards_in_hand]

    sum_val = 0
    for value in values:
        if value == "A":
            sum_val += 11
        elif value in ['J', 'Q', 'K']:
            sum_val += 10
        else:
            sum_val += int(value)

    # correct for Aces
    aces = values.count("A")
    while sum_val > 21 and aces:
        sum_val -= 10
        aces -= 1       # subtract 'aces' after each check

    return sum_val


def compare_cards(player_hand, dealer_hand):
    player_total = total(player_hand)
    dealer_total = total(dealer_hand)

    if not busted(player_hand) and busted(dealer_hand):
        return PLAYER
    elif busted(player_hand) and not busted(dealer_hand):
        return DEALER
    elif player_total > dealer_total:
        return PLAYER
    elif player_total < dealer_total:
        return DEALER
    
    return None


def display_winner(player_hand, dealer_hand):
    winner = compare_cards(player_hand, dealer_hand)

    if not winner:
        print("===================")
        print("ROUND RESULT: TIE!")
        print("===================\n")
    elif winner:
        print("====================================")
        print(f"The winner of this round: {winner}")
        print("====================================\n")


def play_again():
    prompt("Want to play again? Please type (y or n):")
    while True:
        answer = input().strip().lower()

        match answer:
            case 'y':
                return True
            case 'n':
                return False
            case _:
                prompt("Please try either 'y' or 'n':")


def play_game():
    while True:
        terminal_clear()
        prompt("Welcom to Twenty-One!")

        player_hand = deal_cards()
        dealer_hand = deal_cards()

        display_initial_hands(player_hand, dealer_hand)

        player_turn(player_hand)
        dealer_turn(dealer_hand)
        compare_cards(player_hand, dealer_hand)
        display_winner(player_hand, dealer_hand)

        if not play_again():
            break


play_game()

File: js225/lesson6/21.js
const shuffle = require('shuffle-array');
const rlsync = require('readline-sync');

function prompt(mssg) {
  console.log(`===> ${mssg}`);
}

function largeDivider(mssg) {
  console.log('='.repeat(mssg.length));
}

function smallDivider(mssg) {
  console.log('-'.repeat(mssg.length));
}

function pageBreak() {
  console.log("============================================================");
}


class Deck {
  static SUITS = ['Hearts', 'Diamonds', 'Spades', 'Clubs'];
  static VALUES = [
    '2', '3', '4', '5', '6', '7', '8', '9',
    '10', 'Jack', 'Queen', 'King', 'Ace',
  ];

  constructor() {
    this.cards = [];

    Deck.VALUES.forEach(value => {
      Deck.SUITS.forEach(suit => {
        this.cards.push(new Card(value, suit));
      });
    });

    shuffle(this.cards);
  }
}


class Card {
  constructor(value, suit) {
    this.value = value;
    this.suit = suit;
  }

  toString() {
    return `${this.value} of ${this.suit}`;
  }
}


class Hand {
  constructor() {
    this.cards = [];
  }

  toString() {
    return this.cards.map(card => `${card}`).join(', ');
  }

  updateValueTotal() {
    const values = this.cards.map(card => card.value);

    let total = 0;
    values.forEach(value => {
      if (value === 'Ace') total += 11;
      else if (["Jack", "Queen", "King"].includes(value)) total += 10;
      else total += parseInt(value, 10);
    });

    let aces = values.filter(value => value === 'Ace').length;
    while (total > 21 && aces) {
      total -= 10;
      aces -= 1;
    }

    this.valueTotal = total;
  }
}


class Participant {
  constructor() {
    this.hand = new Hand();
  }

  resetHand() {
    this.hand = new Hand();
  }
}


class Player extends Participant {
  constructor() {
    super();
    this.bettingDollars = 5;
  }

  addDollar() {
    this.bettingDollars += 1;
  }

  subtractDollar() {
    this.bettingDollars -= 1;
  }

  isRich() {
    return this.bettingDollars >= 10;
  }

  isPoor() {
    return this.bettingDollars === 0;
  }
}


class Dealer extends Participant {
  constructor() {
    super();
  }
}


class TwentyOneGame {
  static PLAYER = Player.name;
  static DEALER = Dealer.name;
  static BEST_SCORE = 21;
  static DEALER_MIN = 17;

  constructor() {
    this.deck = new Deck();
    this.player = new Player();
    this.dealer = new Dealer();
  }

  start() {
    console.clear();
    this.displayWelcomeMessage();
    this.displayBettingDollarsPurpose();

    while (true) {
      this.displayRemainingPlayerDollars();
      this.dealInitialHands();
      this.displayInitialHands();

      this.processTurns();

      this.compareTotals();
      this.displayFinalTotalAndHands();
      this.displayWinner();
      this.updatePlayerDollars();

      if (this.endGame()) {
        this.displayEndGameMessage();
        break;
      }

      this.resetParticipantHands();
      if (!this.playAgain()) break;
    }

    this.displayGoodbyeMessage();
  }

  dealInitialHands() {
    for (let count = 0; count <= 2; count += 1) {
      this.dealCardTo(this.player);
      this.dealCardTo(this.dealer);
    }
  }

  dealCardTo(participant) {
    /*
    Check if any cards left in deck, intialize deck if not
    Deal cards to each player
    */
    const cardsLeft = this.deck.cards.length;
    if (!cardsLeft) this.deck = new Deck();

    const card = this.deck.cards.pop();
    participant.hand.cards.push(card);
    participant.hand.updateValueTotal();
  }

  displayInitialHands() {
    console.log();
    prompt(`Dealer's hand: ${this.dealer.hand.cards[0]} and ?`);
    prompt(`Your Hand: ${this.player.hand}`);
    console.log();
  }

  getCurrentHand(participant) {
    return participant.hand;
  }

  playerTurn() {
    let playerTotal;

    while (true) {
      playerTotal = this.getHandTotal(this.player);

      if (this.busted(this.player)) break;

      prompt(`PLAYER'S TOTAL: ${playerTotal}`);
      prompt('hit (h) or stay (s)?');
      const answer = rlsync.prompt();

      if (answer === 's') break;
      if (answer === 'h') {
        this.hit(this.player);
        prompt(`PLAYER'S hand: ${this.player.hand}`);
      } else {
        prompt('Try again you silly goose!');
      }
    }

    if (this.busted(this.player)) {
      prompt(`PLAYER'S TOTAL: ${playerTotal}`);
      prompt('You BUSTED!\n');
    } else prompt('You chose to STAY!\n');
  }

  dealerTurn() {
    while (true) {
      const dealerTotal = this.getHandTotal(this.dealer);
      prompt(`DEALER'S hand: ${this.dealer.hand}`);
      prompt(`DEALER'S TOTAL: ${dealerTotal}`);

      if (dealerTotal === TwentyOneGame.BEST_SCORE) {
        prompt('DEALER has 21!\n');
        return;
      }

      if (this.busted(this.dealer)) {
        prompt('DEALER BUSTED!\n');
        break;
      }

      if (this.dealerStays()) {
        prompt('DEALER STAYS.\n');
        break;
      }

      prompt('DEALER HITS!');
      console.log();
      this.hit(this.dealer);
    }
  }

  hit(participant) {
    this.dealCardTo(participant);
  }

  busted(participant) {
    return participant.hand.valueTotal > TwentyOneGame.BEST_SCORE;
  }

  dealerStays() {
    return this.dealer.hand.valueTotal >= TwentyOneGame.DEALER_MIN;
  }

  processTurns() {
    this.playerTurn();

    if (!this.endRound()) {
      this.dealerTurn();
    }
  }

  endRound() {
    return this.busted(this.player);
  }

  compareTotals() {
    const playerTotal = this.getHandTotal(this.player);
    const dealerTotal = this.getHandTotal(this.dealer);

    if (!this.busted(this.player) && this.busted(this.dealer)) TwentyOneGame.PLAYER;
    if (this.busted(this.player) && !this.busted(this.dealer)) TwentyOneGame.DEALER;
    if (playerTotal > dealerTotal) TwentyOneGame.PLAYER;
    if (playerTotal < dealerTotal) TwentyOneGame.DEALER;

    return null;
  }

  getWinner() {
    return this.compareTotals();
  }

  getHandTotal(participant) {
    return participant.hand.valueTotal;
  }

  updatePlayerDollars() {
    const winner = this.getWinner();

    if (!winner) {
      return;
    }

    if (TwentyOneGame.PLAYER.includes(winner)) {
      this.player.addDollar();
    } else {
      this.player.subtractDollar();
    }
  }

  displayFinalTotalAndHands() {
    console.log("ROUND TOTAL & HANDS:\n");
    console.log(`PLAYER:
    Total: ${this.getHandTotal(this.player)}
    Cards: ${this.getCurrentHand(this.player)}`);
    console.log(`DEALER:
    Total: ${this.getHandTotal(this.dealer)}
    Cards: ${this.getCurrentHand(this.dealer)}`);
  }

  displayWinner() {
    const winner = this.getWinner();

    if (!winner) {
      const noWinMsg = "ROUND RESULT: TIE!";
      largeDivider(noWinMsg);
      console.log(noWinMsg);
      largeDivider(noWinMsg);
      console.log()
    } else if (winner) {
      const winMsg = `ROUND WINNER: ${winner}`;
      largeDivider(winMsg);
      console.log(winMsg);
      largeDivider(winMsg);
    }
  }

  displayBettingDollarsPurpose() {
    console.log("If you finish with more than $10, you are RICH! (Game ends)");
    console.log(`If you lose all of your betting dollars, \
you are POOR! (Game ends)`);
  }

  displayRemainingPlayerDollars() {
    console.log(`Remaining BETTING DOLLARS: $${this.player.bettingDollars}`);
  }

  resetParticipantHands() {
    this.player.resetHand();
    this.dealer.resetHand();
  }

  playAgain() {
    prompt('Want to play again? Please type (y or n):');
    let answer;
    while (true) {
      answer = rlsync.prompt()
      answer = answer.trim().toLowerCase();

      switch (answer) {
        case 'y':
          console.clear();
          return true;
        case 'n':
          return false;
        default:
          prompt("Please try either 'y' or 'n': ");
      }
    }
  }

  endGame() {
    if (this.player.isPoor() || this.player.isRich()) true;
    return false;
  }

  displayEndGameMessage() {
    if (this.player.isPoor()) {
      console.log('You are POOR! Game Over!');
    } else if (this.player.isRich()) {
      console.log('You are RICH! Game Over!')
    }
  }

  displayWelcomeMessage() {
    console.log('Welcome to Twenty-One!');
    console.log();
  }

  displayGoodbyeMessage() {
    console.log('Thank you for playing Twenty-One. Goodbye!');
  }
}

const game = new TwentyOneGame();
game.start();